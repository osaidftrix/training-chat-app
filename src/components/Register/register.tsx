import { Link, useNavigate } from 'react-router-dom';
import { getConnection } from '../../store/slices/connectionSlice.js';
import { useAppSelector } from '../../store/hooks.js';
import { useForm, SubmitHandler } from 'react-hook-form';
import { toast } from 'react-toastify';
import { useEffect } from 'react';

type Inputs = {
  userName: string;
  password: string;
};

const Register = () => {
  const { register, handleSubmit } = useForm<Inputs>();
  const navigate = useNavigate();
  const onSubmit: SubmitHandler<Inputs> = (data) => {
    onRegister({ username: data.userName, password: data.password });
  };

  const connection = useAppSelector(getConnection);
  useEffect(() => {
    if (connection.isAuthenticated) {
      navigate('/');
    }
  }, [connection, navigate]);

  const onRegister = async (user: { username: string; password: string }) => {
    if (!connection.connectionInstance) return;
    try {
      const register = await connection.connectionInstance.invoke('Register', user);

      if (register) {
        toast.success('User Registered Successfully!', {
          position: 'top-right',
          autoClose: 200,
          hideProgressBar: true,
          closeOnClick: true,
          theme: 'light',
        });
        navigate('/login');
      } else {
        toast.error('User Already Registered', {
          position: 'top-right',
          autoClose: 200,
          hideProgressBar: true,
          closeOnClick: true,
          theme: 'light',
        });
        return;
      }
    } catch (err) {
      console.log('err', err);
    }
  };

  return (
    <section className="bg-gray-50 ">
      <div className="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0">
        <div className="w-full bg-white rounded-lg shadow  md:mt-0 sm:max-w-md xl:p-0 ">
          <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
            <h1 className="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl ">
              Register to Chat
            </h1>
            <form name="form" role="form" aria-label="form" className="space-y-4 md:space-y-6" onSubmit={handleSubmit(onSubmit)}>
              <div>
                <label htmlFor="text" className="block mb-2 text-sm font-medium text-gray-900 ">
                  Username
                </label>
                <input
                  {...register('userName')}
                  className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5"
                  required
                />
              </div>

              <div>
                <label htmlFor="password" className="block mb-2 text-sm font-medium text-gray-900">
                  Password
                </label>
                <input
                  {...register('password')}
                  type="password"
                  className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5"
                  required
                />
              </div>

              <button
                type="submit"
                className="w-full bg-primary-600 hover:bg-primary-700 focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center"
              >
                Register
              </button>

              <p className="text-sm font-light text-gray-500">
                Already Registered?{' '}
                <Link to={'/login'} className="font-medium text-primary-600 hover:underline">
                  Login Here
                </Link>
              </p>
            </form>
          </div>
        </div>
      </div>
    </section>
  );
};
export default Register;
