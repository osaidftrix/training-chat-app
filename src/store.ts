import { combineReducers, configureStore } from '@reduxjs/toolkit';
import connectionSlice from './store/slices/connectionSlice';
import storage from 'redux-persist/lib/storage';
//import storage from 'redux-persist-indexeddb-storage';
import { persistReducer, persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['connection'],
  // whitelist: ['connection']
};

const rootReducer = combineReducers({
  connection: connectionSlice,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: [thunk],
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
