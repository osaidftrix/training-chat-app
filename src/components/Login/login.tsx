import { useEffect } from 'react';
import { useForm, SubmitHandler } from 'react-hook-form';
import { Link, useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import { getConnection, ConnectionState, setConnection } from '../../store/slices/connectionSlice';

type Inputs = {
  userName: string;
  password: string;
};

const Login = () => {
  const {
    register,
    handleSubmit,
    // watch
    // formState: { errors },
  } = useForm<Inputs>();
  const navigate = useNavigate();
  const onSubmit: SubmitHandler<Inputs> = (data) => {
    onLogin({ username: data.userName, password: data.password });
  };
  const dispatch = useAppDispatch();
  const connection = useAppSelector(getConnection);
useEffect(() => {
    if (connection.isAuthenticated) {
      navigate('/');
    }
  }, [connection, navigate]);

  const onLogin = async (user: { username: string; password: string }) => {
    if (!connection.connectionInstance) return;
    try {
      const login: { allowLogin: boolean; username: null | string } = await connection.connectionInstance.invoke(
        'Login',
        user,
      );
      if (login.allowLogin) {
        toast.success('User Logged In Successfully!', {
          position: 'top-right',
          autoClose: 200,
          hideProgressBar: true,
          closeOnClick: true,
          theme: 'light',
        });
      } else {
        toast.error('Incorrect username or password', {
          position: 'top-right',
          autoClose: 200,
          hideProgressBar: true,
          closeOnClick: true,
          theme: 'light',
        });
        return;
      }
      const connectionPayload: ConnectionState = {
        connectionInstance: connection.connectionInstance,
        connectionId: connection.connectionId,
        isAuthenticated: true,
        userName: user.username,
        loggedInUsers: [],
        messages: { public: [] },
      };
      dispatch(setConnection({ ...connectionPayload }));
    } catch (err) {
      console.log('Could not login: ', err);
    }
  };

  return (
    <section className="bg-gray-50">
      <div className="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0">
        <div className="w-full bg-white rounded-lg shadow md:mt-0 sm:max-w-md xl:p-0">
          <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
            <h1 className="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl">
              Sign in to your account
            </h1>
            <form name="form" role="form" aria-label="form" className="space-y-4 md:space-y-6" onSubmit={handleSubmit(onSubmit)}>
              <div>
                <label htmlFor="text" className="block mb-2 text-sm font-medium text-gray-900">
                  Username
                </label>
                <input
                  {...register('userName')}
                  type="text"
                  className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5"
                  required
                />
              </div>

              <div>
                <label htmlFor="password" className="block mb-2 text-sm font-medium text-gray-900">
                  password
                </label>
                <input
                  {...register('password')}
                  type="password"
                  name="password"
                  className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5"
                  required
                />
              </div>

              <button
                type="submit"
                className="w-full bg-primary-600 hover:bg-primary-700 focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center"
              >
                Sign in
              </button>

              <p className="text-sm font-light text-gray-500">
                Donâ€™t have an account yet?{' '}
                <Link to={'/register'} className="font-medium text-primary-600 hover:underline">
                  Sign Up
                </Link>
              </p>
            </form>
          </div>
        </div>
      </div>
    </section>
  );
};

export default Login;
