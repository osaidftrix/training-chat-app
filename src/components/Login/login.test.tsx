import { describe, test, expect } from 'vitest';
import { MemoryRouter, Routes, Route } from 'react-router';
import Login from './login';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { store, persistor } from '../../store';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Register from '../Register/register';

describe('<Login />', () => {
  test('Login renders correctly', () => {
    const element = render(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <MemoryRouter initialEntries={['/login']}>
            <Routes>
              <Route path="login" element={<Login />}></Route>
            </Routes>
          </MemoryRouter>
        </PersistGate>
      </Provider>,
    );
    expect(element).toBeTruthy();
  });
  test('Login on submit works properly', async () => {
    render(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <MemoryRouter initialEntries={['/login']}>
            <Routes>
              <Route path="login" element={<Login />}></Route>
            </Routes>
          </MemoryRouter>
        </PersistGate>
      </Provider>,
    );
    const buttonElement = screen.getByRole('button');
    await userEvent.click(buttonElement);
    expect(buttonElement).toBeTruthy();
  });
  test('Login redirects to register properly', async () => {
    render(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <MemoryRouter initialEntries={['/login']}>
            <Routes>
              <Route path="login" element={<Login />}></Route>
              <Route path="register" element={<Register />}></Route>
            </Routes>
          </MemoryRouter>
        </PersistGate>
      </Provider>,
    );
    const buttonElement = screen.getByText('Sign Up');
    await userEvent.click(buttonElement);
    const registerHeader = screen.getByText('Register to Chat');
    expect(registerHeader).toBeTruthy();
  });
});
