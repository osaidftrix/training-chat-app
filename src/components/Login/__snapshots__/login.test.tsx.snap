// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Login /> > Login renders correctly 1`] = `
<section
  className="bg-gray-50"
>
  <div
    className="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0"
  >
    <div
      className="w-full bg-white rounded-lg shadow md:mt-0 sm:max-w-md xl:p-0"
    >
      <div
        className="p-6 space-y-4 md:space-y-6 sm:p-8"
      >
        <h1
          className="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl"
        >
          Sign in to your account
        </h1>
        <form
          aria-label="form"
          className="space-y-4 md:space-y-6"
          name="form"
          onSubmit={[Function]}
          role="form"
        >
          <div>
            <label
              className="block mb-2 text-sm font-medium text-gray-900"
              htmlFor="text"
            >
              Username
            </label>
            <input
              className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5"
              name="userName"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="text"
            />
          </div>
          <div>
            <label
              className="block mb-2 text-sm font-medium text-gray-900"
              htmlFor="password"
            >
              password
            </label>
            <input
              className="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="password"
            />
          </div>
          <button
            className="w-full bg-primary-600 hover:bg-primary-700 focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center"
            type="submit"
          >
            Sign in
          </button>
          <p
            className="text-sm font-light text-gray-500"
          >
            Donâ€™t have an account yet?
             
            <a
              className="font-medium text-primary-600 hover:underline"
              href="/register"
              onClick={[Function]}
            >
              Sign Up
            </a>
          </p>
        </form>
      </div>
    </div>
  </div>
</section>
`;
